lib/internal/bootstrap/node.js

```diff
@@ -116,6 +116,18 @@ if (isMainThread) {
   process.exit = wrapped.exit;
 }
 
+{
+  let stillSynchronous = true;
+  Object.defineProperty(process, 'stillSynchronous', {
+    enumerable: true,
+    configurable: false,
+    get() {
+      if (stillSynchronous) stillSynchronous = rawMethods.stillSynchronous();
+      return stillSynchronous;
+    }
```

Comment:
You can use the existing `perf_hooks.performance.nodeTiming.loopStart` timestamp to determine if we're still in the main script or running within the event loop.

```sh
$ node -e "console.log(perf_hooks.performance.nodeTiming.loopStart)"
-1
$ node -e "process.nextTick(() => console.log(perf_hooks.performance.nodeTiming.loopStart))"
-1
$ node -e "setImmediate(() => console.log(perf_hooks.performance.nodeTiming.loopStart))"
32.2972889465332
```